File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/PlaceholderExamples.java
package questpdf.docexamples;

import com.questpdf.fluent.Document;
import com.questpdf.helpers.PageSize;
import com.questpdf.helpers.PageSizes;
import com.questpdf.helpers.Placeholders;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import org.junit.jupiter.api.Test;

import java.util.function.BiConsumer;

public class PlaceholderExamples extends DocExample {

    @Test
    public void textExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(15f);

                            BiConsumer<String, String> addItem = (label, value) -> {
                                column.item().text(text -> {
                                    text.span(label + ": ").bold();
                                    text.span(value);
                                });
                            };

                            addItem.accept("Name", Placeholders.name());
                            addItem.accept("Email", Placeholders.email());
                            addItem.accept("Phone", Placeholders.phoneNumber());
                            addItem.accept("Date", Placeholders.shortDate());
                            addItem.accept("Time", Placeholders.time());
                        });
                });
            })
            .generateImages(index -> output("placeholders-text.webp"), settings);
    }

    @Test
    public void backgroundColorExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(320f, 0f));
                    page.maxSize(new PageSize(320f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .grid(grid -> {
                            grid.columns(5);
                            grid.spacing(5f);

                            for (var i = 0; i < 25; i++) {
                                grid.item()
                                    .height(50f)
                                    .width(50f)
                                    .background(Placeholders.backgroundColor());
                            }
                        });
                });
            })
            .generateImages(index -> output("placeholders-color-background.webp"), settings);
    }

    @Test
    public void colorExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(10f);

                            for (var i = 0; i < 5; i++) {
                                column.item()
                                    .text(Placeholders.sentence())
                                    .fontColor(Placeholders.color());
                            }
                        });
                });
            })
            .generateImages(index -> output("placeholders-color.webp"), settings);
    }

    @Test
    public void imageExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.Best);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .width(200f)
                        .column(column -> {
                            column.spacing(10f);

                            // provide an exact image resolution
                            column.item()
                                .image(Placeholders.image(100, 50));

                            // specify physical width and height of the image
                            // (The C# original passes Placeholders.Image through the Image(Func<ImageSize, byte[]>)
                            // overload, which is not bridged; the payload delegate is the direct equivalent.)
                            column.item()
                                .width(200f)
                                .height(150f)
                                .image(payload -> Placeholders.image(payload.getImageSize()));

                            // specify target physical width and aspect ratio
                            column.item()
                                .width(200f)
                                .aspectRatio(3 / 2f)
                                .image(payload -> Placeholders.image(payload.getImageSize()));
                        });
                });
            })
            .generateImages(index -> output("placeholders-image.webp"), settings);
    }

    @Test
    public void elementExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.High);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.size(PageSizes.getA5());
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.header()
                        .height(100f)
                        .placeholder("Header");

                    page.content()
                        .paddingVertical(25f)
                        .placeholder();

                    page.footer()
                        .height(100f)
                        .placeholder("Footer");
                });
            })
            .generateImages(index -> output("placeholder-element.webp"), settings);
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/PlaceholderExamples.kt
package questpdf.docexamples

import com.questpdf.fluent.Document
import com.questpdf.helpers.PageSize
import com.questpdf.helpers.PageSizes
import com.questpdf.helpers.Placeholders
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import org.junit.jupiter.api.Test

class PlaceholderExamples : DocExample() {

    @Test
    fun textExample() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(15f)

                            fun addItem(label: String, value: String) {
                                item().text {
                                    span("$label: ").bold()
                                    span(value)
                                }
                            }

                            addItem("Name", Placeholders.name())
                            addItem("Email", Placeholders.email())
                            addItem("Phone", Placeholders.phoneNumber())
                            addItem("Date", Placeholders.shortDate())
                            addItem("Time", Placeholders.time())
                        }
                }
            }
            .generateImages({ output("placeholders-text.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun backgroundColorExample() {
        Document
            .create {
                page {
                    minSize(PageSize(320f, 0f))
                    maxSize(PageSize(320f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .grid {
                            columns(5)
                            spacing(5f)

                            repeat(25) {
                                item()
                                    .height(50f)
                                    .width(50f)
                                    .background(Placeholders.backgroundColor())
                            }
                        }
                }
            }
            .generateImages({ output("placeholders-color-background.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun colorExample() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(10f)

                            repeat(5) {
                                item()
                                    .text(Placeholders.sentence())
                                    .fontColor(Placeholders.color())
                            }
                        }
                }
            }
            .generateImages({ output("placeholders-color.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun imageExample() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .width(200f)
                        .column {
                            spacing(10f)

                            // provide an exact image resolution
                            item()
                                .image(Placeholders.image(100, 50))

                            // specify physical width and height of the image
                            // (The C# original passes Placeholders.Image through the Image(Func<ImageSize, byte[]>)
                            // overload, which is not bridged; the payload delegate is the direct equivalent.)
                            item()
                                .width(200f)
                                .height(150f)
                                .image { payload -> Placeholders.image(payload.imageSize) }

                            // specify target physical width and aspect ratio
                            item()
                                .width(200f)
                                .aspectRatio(3 / 2f)
                                .image { payload -> Placeholders.image(payload.imageSize) }
                        }
                }
            }
            .generateImages({ output("placeholders-image.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.Best
                rasterDpi = 144
            })
    }

    @Test
    fun elementExample() {
        Document
            .create {
                page {
                    size(PageSizes.A5)
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    header()
                        .height(100f)
                        .placeholder("Header")

                    content()
                        .paddingVertical(25f)
                        .placeholder()

                    footer()
                        .height(100f)
                        .placeholder("Footer")
                }
            }
            .generateImages({ output("placeholder-element.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.High
                rasterDpi = 144
            })
    }
}
